include/circt/Dialect/ESI/ESIChannels.td

```diff
@@ -197,6 +197,78 @@ def UnwrapSVInterfaceOp : ESI_Op<"unwrap.iface", [
   let hasVerifier = 1;
 }
 
+//===----------------------------------------------------------------------===//
+// Channel bundles
+//===----------------------------------------------------------------------===//
+
+// Used to indicate the presence of a keyword in asm. Should be used within
+// optional parens in the declarative assembly format.
+// Upstreaming: https://github.com/llvm/llvm-project/pull/65438
+def UnitParameter : AttrOrTypeParameter<"::mlir::UnitAttr", "boolean flag"> {
+  let printer = "";
+  let parser = "::mlir::UnitAttr::get($_parser.getContext())";
+  let defaultValue = "::mlir::UnitAttr()";
+}
+
+def ChannelDirection : I32EnumAttr<"ChannelDirection",
+  "Direction of channel", [
+    I32EnumAttrCase<"to", 1>,
+    I32EnumAttrCase<"from", 2>,
+  ]>;
+
+def ChannelBundleImpl : ESI_Type<"ChannelBundle"> {
+  let summary = "a group of related channels";
+
+  let description = [{
+    A channel bundle (sometimes referred to as just "bundle") is a set of
+    channels of associated signals, along with per-channel names and directions.
+    The prototypical example for a bundle is a request-response channel pair.
+
+    The direction terminology is a bit confusing. Let us designate the module
+    which is outputting the bundle as the "sender" module and a module which has
+    a bundle as an input as the "receiver". The directions "from" and "to" are
+    from the senders perspective. So, the "to" direction means that channel is
+    transmitting messages from the sender to the receiver. Then, "from" means
+    that the sender is getting messages from the receiver (typically responses).
+
+    When requesting a bundle from a service, the client is always considered the
+    sender. So the "to" direction is for the client to send messages to the
+    service.
```

Comment:
I think it would be prudent to weave this kind of terminology description into the ESI rationale.

Reply:
I'll need to update all of the non-ODS ESI documentation at some point based on the internal document we went over. As of now, the rationale doesn't even mention bundles. So "prudent" is accurate.